src/blob/services/blob.service.ts

```diff
@@ -0,0 +1,60 @@
+import { mkdir, readFile, rm, writeFile } from "node:fs/promises";
+import { createReadStream, existsSync } from "node:fs";
+import mime from "mime-types";
+import { config } from "../../config.js";
+import { getFullFileDir } from "../utils/filesystem.js";
+import { extractRawContent } from "../helpers.js";
+import type { BlobMetadata } from "../types.js";
+
+export class BlobService {
+  static async getBlob(id: string) {
+    const fullFilePath = `${await getFullFileDir(config.BLOBS_DIR, id)}/${id}`;
+
+    if (!existsSync(fullFilePath)) {
+      return null;
+    }
+
+    const readStream = createReadStream(fullFilePath);
+    const metadataContent = await readFile(
+      `${config.METADATA_DIR}/${id}.metadata`
+    );
+    const metadata: BlobMetadata = JSON.parse(
+      metadataContent.toString("ascii")
+    );
+
+    const contentType =
+      metadata.headers?.["content-type"] ??
+      (mime.lookup(fullFilePath) || "application/octet-stream");
+
+    return {
+      stream: readStream,
+      headers: metadata.headers,
+      contentType,
+    };
+  }
+
+  static async createBlob(id: string, request: any, headers: Buffer) {
+    const fileDir = await getFullFileDir(config.BLOBS_DIR, id);
```

Comment:
I'm not sure it's correct:
the function `getFullFileDir` depends on the amount of files in a directory, and selects/generates the relevant directory.
let's examine the following scenario:
POST /blobs/xyz.png
say hash is "abcdef"
say storage/ab is full
so we check storage/ab/cd - and it has enough space. cool, we store the blob in storage/ab/cd/xyz.png
now let's delete a few files:
DELETE /blobs/foo1.png
DELETE /blobs/foo2.png
DELETE /blobs/foo3.png
let's assume that they were all stored in storage/ab
now again POST /blobs/xyz.png
since storage/ab is not full, we'll store it twice (!)